xmake/rules/vala/xmake.lua

```diff
@@ -103,22 +110,39 @@ rule("vala.build")
                 table.insert(argv, path(headerfile))
             end
         end
+
         local vapidir = target:data("vala.vapidir")
         if vapidir then
             table.insert(argv, path(vapidir, function (p) return "--vapidir=" .. p end))
         end
+
         local valaflags = target:data("vala.flags")
         if valaflags then
             table.join2(argv, valaflags)
         end
-        table.insert(argv, path(sourcefile_vala))
+
+        batchcmds:show_progress(opt.progress, "${color.build.object}compiling.vala")
         batchcmds:vrunv(valac.program, argv)
-        batchcmds:compile(sourcefile_c, objectfile)
+    end)
+
+    on_buildcmd_file(function (target, batchcmds, sourcefile_vala, opt)
+        -- Again, only vala files need special treatment
+        if string.match(sourcefile_vala, ".vala$") ~= nil then
+            local sourcefile_c = target:autogenfile((sourcefile_vala:gsub(".vala$", ".c")))
+            local basedir = path.directory(sourcefile_c)
+
+            batchcmds:mkdir(basedir)
 
-        -- add deps
-        batchcmds:add_depfiles(sourcefile_vala)
-        batchcmds:set_depmtime(os.mtime(objectfile))
-        batchcmds:set_depcache(target:dependfile(objectfile))
+            local objectfile = target:objectfile(sourcefile_c)
+            table.insert(target:objectfiles(), objectfile)
+
+            batchcmds:add_depfiles(sourcefile_vala)
+            batchcmds:set_depmtime(os.mtime(objectfile))
+            batchcmds:set_depcache(target:dependfile(objectfile))
+
+            batchcmds:show_progress(opt.progress, "${color.build.object}compiling.c %s", sourcefile_c)
+            batchcmds:compile(sourcefile_c, objectfile, { configs = { force = { cflags = "-w" } } })
```

Comment:
why add `-w`? we can use `set_warnings` in project xmake.lua

Reply:
Because there's expected warnings from files that are generated from Vala and it's [recommended to disable them by Vala devs](http://docs.vala.dev/tutorials/programming-language/main/07-00-tools/07-01-valac.html):
```
The recommended approach is to just disable those warnings by passing options to the C compiler:
valac -X -w hello.vala
```
And this option only going to touch valac generated C files